### PR TITLE
[SPARK-50322][SQL] Fix parameterized identifier in a sub-query

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to postpone parameters resolution till `UnresolvedWithCTERelations` is resolved. 

### Why are the changes needed?
To fix the query failure:
```sql
execute immediate 'with v1 as (select * from tt1 where 1 = (Select * from identifier(:tab))) select * from v1' using 'tt1' as tab;
[UNBOUND_SQL_PARAMETER] Found the unbound parameter: tab. Please, fix `args` and provide a mapping of the parameter to either a SQL literal or collection constructor functions such as `map()`, `array()`, `struct()`. SQLSTATE: 42P02
```

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running new test:
```
$ build/sbt "sql/test:testOnly org.apache.spark.sql.ParametersSuite
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by srielau — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala [hunk 1]
```diff
@@ -741,4 +741,21 @@ class ParametersSuite extends QueryTest with SharedSparkSession with PlanTest {
...
+
+  test("SPARK-50322: parameterized identifier in a sub-query") {
+    withTable("tt1") {
+      sql("create table tt1(c1 int)")
+      sql("insert into tt1 values (1)")
+      def query(p: String): String = {
+        s"""
+          |with v1 as (
+          |  select * from tt1
+          |  where 1 = (Select * from identifier($p))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:753]:
```suggestion
          |  where 1 = (select * from identifier($p))
```

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala:753]: or we should upper case all SQL keywords

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/ParametersSuite.scala]: I uppercased all SQL keywords in the test.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 2]
```diff
@@ -189,7 +189,8 @@ object BindParameters extends ParameterizedQueryProcessor with QueryErrorsBase {
       // We should wait for `CTESubstitution` to resolve CTE before binding parameters, as CTE
       // relations are not children of `UnresolvedWith`.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:190]: I'm thinking of a different approach: shall we explicitly match `UnresolvedWith` here and bind parameters in the CTE relations? Then we no longer need this hack to delay parameter binding.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:190]: Can we do that in 2 steps: 1. fix the issue, 2. do the proposed refactoring? It seems less risky, so, the first patch could be backported more safely.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:190]: +1

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srielau and @cloud-fan for review.

**Comment by MaxGekk** [comment]
We don't need to backport this fix to `branch-3.5` because it doesn't have the changes https://github.com/apache/spark/pull/47180.

**Pull request closed without merging**